File: vc-go/status/status.go
// Package status decodes and encodes Bitstring Status List 1.0 lists.
package status

import (
	"bytes"
	"compress/gzip"
	"encoding/base64"
	"errors"
	"fmt"
	"io"
	"strings"
)

const (
	// ListBits is the number of bits in a Vemphy status list.
	ListBits  = 131_072
	listBytes = ListBits / 8
)

var (
	ErrEncoding   = errors.New("status list is not multibase base64url around gzip")
	ErrSize       = fmt.Errorf("status list must be %d bytes", listBytes)
	ErrOutOfRange = errors.New("status index is outside the list")
)

// Decode reads an encodedList: multibase base64url (the letter u) around
// gzip. Decompression stops as soon as the output passes the expected size,
// so a small input cannot be used to exhaust memory.
func Decode(encodedList string) ([]byte, error) {
	rest, ok := strings.CutPrefix(encodedList, "u")
	if !ok {
		return nil, ErrEncoding
	}
	compressed, err := base64.RawURLEncoding.DecodeString(rest)
	if err != nil {
		return nil, ErrEncoding
	}
	reader, err := gzip.NewReader(bytes.NewReader(compressed))
	if err != nil {
		return nil, ErrEncoding
	}
	bits, err := io.ReadAll(io.LimitReader(reader, listBytes+1))
	if err != nil {
		return nil, ErrEncoding
	}
	if len(bits) != listBytes {
		return nil, ErrSize
	}
	return bits, nil
}

// Encode is deterministic: the gzip header carries no timestamp or file name.
func Encode(bits []byte) (string, error) {
	if len(bits) != listBytes {
		return "", ErrSize
	}
	var buf bytes.Buffer
	w, _ := gzip.NewWriterLevel(&buf, gzip.BestCompression)
	if _, err := w.Write(bits); err != nil {
		return "", err
	}
	if err := w.Close(); err != nil {
		return "", err
	}
	return "u" + base64.RawURLEncoding.EncodeToString(buf.Bytes()), nil
}

// Get reports whether a bit is set. Bit 0 is the most significant bit of the first byte.
func Get(bits []byte, index int) (bool, error) {
	if index < 0 || index >= len(bits)*8 {
		return false, ErrOutOfRange
	}
	return bits[index>>3]&(0x80>>(index&7)) != 0, nil
}

func Set(bits []byte, index int) error {
	if index < 0 || index >= len(bits)*8 {
		return ErrOutOfRange
	}
	bits[index>>3] |= 0x80 >> (index & 7)
	return nil
}

File: packages/vc/src/status.ts
import { base64urlnopad } from '@scure/base'
import { LIST_BITS } from './constants.js'

export { LIST_BITS }

const LIST_BYTES = LIST_BITS / 8

/**
 * Decodes an `encodedList`: multibase base64url (the letter u) around gzip.
 * Decompression stops as soon as the output passes the expected size, so a
 * small input cannot be used to exhaust memory.
 */
export async function decodeList(encodedList: string): Promise<Uint8Array> {
  if (!encodedList.startsWith('u')) throw new Error('status list is not multibase base64url')
  const compressed = base64urlnopad.decode(encodedList.slice(1))
  const bits = await gunzip(compressed, LIST_BYTES)
  if (bits.length !== LIST_BYTES) throw new Error(`status list is ${bits.length} bytes, expected ${LIST_BYTES}`)
  return bits
}

export async function encodeList(bits: Uint8Array): Promise<string> {
  if (bits.length !== LIST_BYTES) throw new Error(`status list must be ${LIST_BYTES} bytes`)
  const compressed = await collect(blobStream(bits).pipeThrough(new CompressionStream('gzip')), Infinity)
  // The gzip header records a timestamp and the operating system. Both are
  // cleared so that the same list always encodes to the same text.
  compressed.fill(0, 4, 9)
  compressed[9] = 0xff
  return 'u' + base64urlnopad.encode(compressed)
}

/** Bit 0 is the most significant bit of the first byte. */
export function getBit(bits: Uint8Array, index: number): boolean {
  return (bits[byteOf(bits, index)]! & (0x80 >> (index & 7))) !== 0
}

export function setBit(bits: Uint8Array, index: number): void {
  bits[byteOf(bits, index)]! |= 0x80 >> (index & 7)
}

function byteOf(bits: Uint8Array, index: number): number {
  if (!Number.isInteger(index) || index < 0 || index >= bits.length * 8) {
    throw new RangeError(`status index ${index} is outside the list`)
  }
  return index >> 3
}

function blobStream(bytes: Uint8Array): ReadableStream<Uint8Array> {
  return new Blob([bytes as Uint8Array<ArrayBuffer>]).stream()
}

function gunzip(bytes: Uint8Array, limit: number): Promise<Uint8Array> {
  return collect(blobStream(bytes).pipeThrough(new DecompressionStream('gzip')), limit)
}

async function collect(stream: ReadableStream<Uint8Array>, limit: number): Promise<Uint8Array> {
  const reader = stream.getReader()
  const chunks: Uint8Array[] = []
  let total = 0
  for (;;) {
    const { done, value } = await reader.read()
    if (done) break
    total += value.length
    if (total > limit) {
      await reader.cancel()
      throw new Error('status list is larger than expected')
    }
    chunks.push(value)
  }
  const out = new Uint8Array(total)
  let offset = 0
  for (const chunk of chunks) {
    out.set(chunk, offset)
    offset += chunk.length
  }
  return out
}
